### ISSUE TITLE
When it's safe to free lws_write buffer

### ISSUE THREAD

**Issue by qdztxc** [warmcat/libwebsockets#2563]
Hi Andy,

I'm using lws as http client.
I call lws_write with buffer allocated from heap. Such as
char *buffer = (char*)malloc(LWS_PRE + sizeof(body));
memcpy(buffer + LWS_PRE, body, body_len);
lws_write(wsi, (unsigned char *)buffer[LWS_PRE],  body_len, LWS_WRITE_HTTP);

My question is after lws_write, when it's safe to free buffer memory? 
Is it copied to lws struct (since lws is async, I think it's required). If it's copied, is it better to just pass raw data (body) without LWS_PRE? lws could allocate internal buffer with LWS_PRE + body size then copy  body to the internal buffer.

Thanks and regards

**Comment by lws-team** [comment]
You can always have the buffer go out of scope after the return from lws_write().

The reasons you can do that are 1) you cannot use lws_write() until you were told it was WRITEABLE, what it means is the OS has already indicated that it will accept some data on the socket, so we know we are "welcome".  And 2) lws will take responsibility for anything rejected by stashing it on heap and sending it in the background, while concealing the writeability while it is being hijacked for that.

Basically the async-ness is mostly moved to before you do the write, rather than after.

The OS decides how much data it will accept before truncating it.  So if you shove large writes in one hit on lws, it will cope with it, but it is inefficient.  Most OS will always accept something like 2KB or 4KB on the socket but larger things may have to be buffered and truncated on heap automatically.  It's better to issue medium size chunks iteratively, asking to write again while there is more.
